psi4/src/psi4/libfock/v.cc

```diff
@@ -400,6 +438,124 @@ double VBase::vv10_nlc(SharedMatrix D, SharedMatrix ret) {
+    for (size_t i = 0; i < num_threads_; i++) {
+        G_local.push_back(std::make_shared<Matrix>("G Temp", natom, 3));
+        U_local.push_back(std::make_shared<Matrix>("U Temp", max_points, max_functions));
+    }
+
+// => Compute the kernel <=
+#pragma omp parallel for private(rank) schedule(guided) num_threads(num_threads_)
+    for (size_t Q = 0; Q < nlgrid.blocks().size(); Q++) {
+// Get thread info
+#ifdef _OPENMP
+        rank = omp_get_thread_num();
+#endif
+
+        // Get per rank-workers
+        std::shared_ptr<BlockOPoints> block = nlgrid.blocks()[Q];
+        std::shared_ptr<SuperFunctional> fworker = functional_workers_[rank];
+        std::shared_ptr<PointFunctions> pworker = nl_point_workers[rank];
+        const std::vector<int>& function_map = block->functions_local_to_global();
+        const int nlocal = function_map.size();
+        const int npoints = block->npoints();
+        double** Tp = pworker->scratch()[0]->pointer();
+
+        // Compute Rho, Phi, etc
+        pworker->compute_points(block);
+
+        // Updates the vals map and returns the energy
+        std::map<std::string, SharedVector> vals = fworker->values();
+
+        parallel_timer_on("Kernel", rank);
+        vv10_exc[rank] += fworker->compute_vv10_kernel(pworker->point_values(), vv10_cache, block, npoints, true);
+        parallel_timer_off("Kernel", rank);
+
+        parallel_timer_on("V_xc gradient", rank);
+
+        // => LSDA and GGA gradient contributions <= //
+        dft_integrators::rks_gradient_integrator(primary_, block, fworker, pworker, G_local[rank], U_local[rank]);
+
+        // => Grid gradient contributions <= //
+        double** Gp = G_local[rank]->pointer();
+        const double* x_grid = fworker->vv_value("GRID_WX")->pointer();
+        const double* y_grid = fworker->vv_value("GRID_WY")->pointer();
+        const double* z_grid = fworker->vv_value("GRID_WZ")->pointer();
+        double** phi = pworker->basis_value("PHI")->pointer();
+        double** phi_x = pworker->basis_value("PHI_X")->pointer();
+        double** phi_y = pworker->basis_value("PHI_Y")->pointer();
+        double** phi_z = pworker->basis_value("PHI_Z")->pointer();
+
+        // double xcontrib = 0.0;
+        // double ycontrib = 0.0;
+        // double zcontrib = 0.0;
+        // for (size_t i = 0; i < npoints; i++){
+        //     xcontrib += x_grid[i];
+        //     ycontrib += y_grid[i];
+        //     zcontrib += z_grid[i];
+        // }
+        for (int P = 0; P < npoints; P++) {
+            std::fill(Tp[P], Tp[P] + nlocal, 0.0);
+            C_DAXPY(nlocal, z_grid[P], phi[P], 1, Tp[P], 1);
+        }
+        for (int ml = 0; ml < nlocal; ml++) {
+            int A = primary_->function_to_center(function_map[ml]);
+            // Gp[A][0] += C_DDOT(npoints, &Tp[0][ml], max_functions, &phi_x[0][ml], max_functions);
+            // Gp[A][1] += C_DDOT(npoints, &Tp[0][ml], max_functions, &phi_y[0][ml], max_functions);
```

Comment:
i'm guessing the x and y terms are to be uncommented eventually?

Reply:
Yea, it is all wildly incorrect until we can isolate blocks to single basis functions. Noted this in the code.